### PR TITLE
Immer support

### PR DESCRIPTION
**Pull request by Aslemammad**
#11 It's ready I think, Feel free to leave comments, suggestions, and ....

### CODE REVIEW

**Comment by dai-shi** [comment]
Hey, nice work.
Are you able to run `yarn build` successfully locally?

File: src/immer/withImmer.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable import/named */
+import { produce, Draft } from 'immer'
+import { atom, WritableAtom } from 'jotai'
+
+export const withImmer = <Value>(anAtom: WritableAtom<Value, Value>) =>
```
Comment (dai-shi) [inline src/immer/withImmer.ts:5]: In the future, we may try using different types for Update: `WritableAtom<ValueToRead, ValueToWrite>`

File: src/immer/atomWithImmer.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable import/named */
+import { produce, Draft } from 'immer'
+import { atom, WritableAtom } from 'jotai'
+
+export const atomWithImmer = <Value>(initialValue: Value) => {
+  const anAtom: WritableAtom<
+    Value,
+    Function | ((draft: Draft<Value>) => void)
+  > = atom(initialValue, (get, set, fn) =>
+    set(anAtom, produce(get(anAtom), (draft) => fn(draft)) as Function)
```
Comment (dai-shi) [inline src/immer/atomWithImmer.ts:10]: When I tried in codesandbox, it seemed to be working without `as Function`. Not sure. ts version could be different. Not super important either.

**Comment by Aslemammad** [comment]
> Hey, nice work.
> Are you able to run `yarn build` successfully locally?

For the first time, Yes, but I deleted the node_modlues and install it, Then it shows me immer not found error as I expected.

**Comment by dai-shi** [comment]
@Aslemammad Try adding it in devDependencies and see how it goes.

File: src/immer/withImmer.ts (hunk 1 above)
Comment (Aslemammad) [reply src/immer/withImmer.ts:5]: I found the solution for this and [this](https://github.com/pmndrs/jotai/pull/120#pullrequestreview-505628651), it gave me error at first, but now, I converted them to unknown then to the Update type. [codesandbox](https://codesandbox.io/s/jotai-immer-typescript-7s1fp?file=/src/App.tsx)

File: src/immer/atomWithImmer.ts (hunk 2 above)
Comment (Aslemammad) [reply src/immer/atomWithImmer.ts:10]: > @Aslemammad Try adding it in devDependencies and see how it goes.

It's OK now, but wait, I should add the last updates for [better types](https://github.com/pmndrs/jotai/pull/120#discussion_r502401629) .

**Comment by Aslemammad** [comment]
> @Aslemammad Try adding it in devDependencies and see how it goes.

It's OK now, but wait, I should add the last updates for better types .

File: src/immer/withImmer.ts (hunk 1 above)
Comment (dai-shi) [reply src/immer/withImmer.ts:5]: No, `as unknown` is not nice for this one. Let's just leave it.

File: src/immer/atomWithImmer.ts (hunk 2 above)
Comment (dai-shi) [reply src/immer/atomWithImmer.ts:10]: https://github.com/pmndrs/jotai/pull/120#discussion_r502401629
For this one, `as unknown` or `as any` is unavoidable. But, `as Function` looks good. Let's keep it.

**Comment by dai-shi** [comment]
> It's OK now

Please push it and see if CI works.

For types, no need to modify as commented.

**Comment by Aslemammad** [comment]
@dai-shi How now?

**Comment by dai-shi** [comment]
Please revert e7813611be6d48ec79b0ff8ce82fe8d986da3441

It seems like you need to run `yarn`.

**Comment by dai-shi** [comment]
We need add the following in "copy" script in package.json:
`shx cp dist/immer.d.ts dist/immer.cjs.d.ts`
put it right after the same one with utils.

**Comment by Aslemammad** [comment]
> Please revert [e781361](https://github.com/pmndrs/jotai/commit/e7813611be6d48ec79b0ff8ce82fe8d986da3441)
> 
> It seems like you need to run `yarn`.

Hmm, Why?

**Comment by dai-shi** [comment]
Look at this: https://codesandbox.io/s/jotai-immer-typescript-forked-z6s2k?file=/src/App.tsx
It's working good!

**Comment by dai-shi** [comment]
> Hmm, Why?

https://github.com/pmndrs/jotai/pull/120/checks?check_run_id=1231379912

```
error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by Aslemammad** [comment]
@dai-shi I ran the yarn command, what now? What is the problem.

**Comment by Aslemammad** [comment]
@dai-shi I solved it, Sorry, it was a dumb problem. It's OK now.

**Comment by Aslemammad** [comment]
@dai-shi I'm writing the tests because I've got the time and I want to spend the time on it.

File: src/immer/withImmer.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable import/named */
+import { produce, Draft } from 'immer'
+import { atom, WritableAtom } from 'jotai'
+
+export const withImmer = <Value, Update>(anAtom: WritableAtom<Value, Update>) =>
```
Comment (dai-shi) [inline src/immer/withImmer.ts:5]:
```suggestion
export const withImmer = <Value>(anAtom: WritableAtom<Value, Value>) =>
```

Sorry to bother you. My suggestion was too naive. Please revert back to this.

**Review by dai-shi — CHANGES_REQUESTED** [review]
One minor request for change

**Comment by dai-shi** [comment]
@Aslemammad Tests would be nice. Docs are more important. I can merge this PR without them, but would like you to work on both if you are interested.

File: src/immer/withImmer.ts (hunk 3 above)
Comment (Aslemammad) [reply src/immer/withImmer.ts:5]: No, I'm learning and i love suggestions, I'm gonna revert.

**Comment by Aslemammad** [comment]
> @Aslemammad Tests would be nice. Docs are more important. I can merge this PR without them, but would like you to work on both if you are interested.

Aha, I'm gonna work on both of them.

**Comment by Aslemammad** [comment]
@dai-shi Done.

**Review by dai-shi — APPROVED** [review]
Thanks!

**Comment by dai-shi** [comment]
Created a follow-up task for you: #122

**Pull request merged**